### ISSUE TITLE
Allow specifying primary key in iteration by secondary key

### ISSUE THREAD

**Issue by knazarov** [tarantool/tarantool#3898]
We have a large space with a secondary index containing many non-unique elements. In order to start iterating by secondary key from a specific tuple, we have to first do `index:pairs(secondary-value)` and then sequentially skip a certain (potentially large) number of tuples.

This problem presented itself when we tried to implement a generic solution for pagination. When paginating by secondary indexes our current implementation has large overhead.

I suggest to add an ability to specify primary index value when iterating by secondary index. Something like:

```lua
index:pairs(key, iterator_type, [primary_key])
```

**Comment by Totktonada** [comment]
I met the same problem when implementing cursors for chunked data transfer for tarantool/graphql. Pagination/cursors on the secondary non-unique indexes can be implemented (using `count` parameter), but it is surprisingly tricky (and so error-prone) when trying to cover all cases (to be honest, I shy to share my current code for that and ask to believe me). Moreover, using `count` is not optimal.

So I strongly vote to implement this!

When I first think about that I imagine that the API can allow to add primary key parts at and of the secondary key ones or the API where they are provided via different parameters (like Konstantin writes above). I don't think it is matter. Maybe separate params are more safe in case of alter / migration.

**Comment by knazarov** [comment]
Actually, adding primary key parts at the end of fully specified secondary key looks good. It doesn't break current API and I can't find any downsides.

**Comment by alyapunov** [comment]
@knazarov Could you just add primary key fields to the secondary index definition and make it unique?

**Comment by knazarov** [comment]
I personally can work around this, but we are planning to give our customers an API  for querying objects by primary/secondary keys across the cluster. They will have problems in understanding why they have to add parts of primary key to each of their secondary keys. So, I doubt this is a good solution.

**Comment by Gerold103** [comment]
If you want to search by secondary + primary key, you need to specify secondary + primary key. Both in index definition and in selects. I don't see any problems here. Just a normal index definition and usual keys. On the contrary, an ability to add a not specified primary key for a secondary index search, which **does not have the primary parts explicitly specified**, looks extremely confusing.

**Comment by knazarov** [comment]
The problem here is for the user: in order to have pagination by secondary index, they have to think about adding primary key parts there. We already have stable iterators that make use of the fact that there are implicit primary index parts in every secondary index. What I want is just for this API to be available for "remote" usage.

**Comment by alyapunov** [comment]
I made it but actually I don't like it.
<details>
<summary>Obvious case</summary>

```
tarantool> s = box.schema.space.create('space')
---
...

tarantool> i1 = s:create_index('primary', {parts={1, 'unsigned'}})
---
...

tarantool> i2 = s:create_index('secondary', {parts={2, 'unsigned'},unique=false})
---
...

tarantool> for i = 1,5 do for j = 1,5 do s:replace{i * 5 + j - 5, j} end end
---
...

tarantool> i2:select{3}
---
- - [3, 3]
  - [8, 3]
  - [13, 3]
  - [18, 3]
  - [23, 3]
...

tarantool> i2:select{3, 15}
---
- []
...

tarantool> i2:select({3, 15}, {iterator='GE', limit=3})
---
- - [18, 3]
  - [23, 3]
  - [4, 4]
...
```
</details>
<details>
<summary>Not so obvious case</summary>

```
tarantool> s = box.schema.space.create('space')
---
...

tarantool> i1 = s:create_index('primary', {parts={{1, 'unsigned'}, {2, 'unsigned'}}})
---
...

tarantool> i2 = s:create_index('secondary', {parts={{1, 'unsigned'}, {3, 'unsigned'}}, unique=false})
---
...

tarantool> s:replace{1, 1, 2}
---
- [1, 1, 2]
...

tarantool> s:replace{1, 2, 2}
---
- [1, 2, 2]
...

tarantool> s:replace{1, 3, 1}
---
- [1, 3, 1]
...

tarantool> s:replace{1, 4, 3}
---
- [1, 4, 3]
...

tarantool> i2:select{1, 2}
---
- - [1, 1, 2]
  - [1, 2, 2]
...

tarantool> i2:select{1, 2, 2}
---
- - [1, 2, 2]
...
```
</details>

**Comment by olegrok** [comment]
> Not so obvious case

```lua
key_def_lib = require('key_def')
pk_key_def = key_def_lib.new(i1.parts)
sk_key_def = key_def_lib.new(i2.parts)
sk_cmp_def = key_def_lib.merge(sk_key_def, pk_key_def)
tarantool> sk_cmp_def 
---
- - type: unsigned
    is_nullable: false
    fieldno: 1
  - type: unsigned
    is_nullable: false
    fieldno: 3
  - type: unsigned
    is_nullable: false
    fieldno: 2
...
cursor = sk_cmp_def:extract_key(tuple)
```

This way could be easily used to extract "cursor" from tuple. But I agree. It looks strange.
Another way is to do merge of keys under the hood:

```lua
i2:select({1, 1}, {key_extension = {1, 2}})
```

But seems it won't high performance solution.

**Comment by alyapunov** [comment]
I made it that way: 
```
i2:select({1, 1}, {key_extension = {1, 2}}}
i2:pairs({1, 1}, {key_extension = {1, 2}}}
```
Of course that's not so optimal as adding parts directly to the key, but both the methods work so you can choose.

**Comment by Mons** [comment]
Why not just create _correct_ index for such task: index which will contain pk as last part of key?

**Comment by alyapunov** [comment]
> 
> 
> Why not just create _correct_ index for such task: index which will contain pk as last part of key?

Please see the discussion above

**Comment by Totktonada** [comment]
I'll try to summarize discussions here and offline ones.

The problem: we have non-unique indices that're technically ready to implement pagination on top of it, but current APIs does not allow it.

There are different ways to solve it.

1. The way proposed here: expose the fact that primary key parts are implicitly added to secondary key parts to :select() API.

   Downside: It looks as leak of implementation details into the high level API.

   Downside: When one want to write some universal pagination code, (s)he need to extract a secondary and primary keys from a tuple using key_def module (and pass them merged or separately to select()).

   (Downside?) Now altering index with addition of new key parts into the end will not change behaviour of existing select requests by this index.

2. Allow to use a tuple instead of a key for select (@alyapunov).

3. Allow to (de)serialize a space iterator (@Mons).

4. Implement server-side cursor.

**Comment by Totktonada** [comment]
How about implementing the hack (1) under `box.internal` namespace to unblock the Solution Team and think about serializable / resumable iterators (3) within another issue?

Possible ways to implement (1):

1.1. Allow to add primary key parts (which are not part of a secondary key) to select / pairs key (those select / pairs, which are under `box.internal`).
1.2. Provide a way to pass a primary key in addition to a secondary key (when all key parts are provided).
1.3. Pass a tuple as a key (the idea is from @alyapunov).

`1.3` is most simple for a user to implement pagination using a GT iterator.

**Comment by alyapunov** [comment]
I made a kind of hack. If you run the code after box.cfg{}:

<details>
<summary>In Tarantool 1.10.7-0-g554d439</summary>

```lua
ffi = require('ffi')

ffi.cdef(
[[
typedef struct tuple_format box_tuple_format_t;
typedef struct tuple box_tuple_t;

box_tuple_format_t *
box_tuple_format(box_tuple_t *tuple);

box_tuple_t *
box_tuple_new(box_tuple_format_t *format, const char *data,
              const char *end);

int
box_tuple_ref(box_tuple_t *tuple);
void
box_tuple_unref(box_tuple_t *tuple);

struct iterator {
	/**
	 * Iterate to the next tuple.
	 * The tuple is returned in @ret (NULL if EOF).
	 * Returns 0 on success, -1 on error.
	 */
	int (*next)(struct iterator *it, struct tuple **ret);
	/** Destroy the iterator. */
	void (*free)(struct iterator *);
	/** Space cache version at the time of the last index lookup. */
	uint32_t space_cache_version;
	/** ID of the space the iterator is for. */
	uint32_t space_id;
	/** ID of the index the iterator is for. */
	uint32_t index_id;
	/**
	 * Pointer to the index the iterator is for.
	 * Guaranteed to be valid only if the schema
	 * state has not changed since the last lookup.
	 */
	struct index *index;
};

struct matras_view {
    /* root extent of the view */
    void *root;
    /* block count in the view */
    uint32_t block_count;
    /* all views are linked into doubly linked list */
    struct matras_view *prev_view, *next_view;
};

struct memtx_tree_iterator {
    /* ID of a block, containing element. -1 for an invalid iterator */
    uint32_t block_id;
    /* Position of an element in the block. Could be -1 for last in block*/
    uint16_t pos;
    /* Version of matras memory for MVCC */
    struct matras_view view;
};

struct memtx_tree_key_data
{
	/** Sequence of msgpacked search fields */
	const char *key;
	/** Number of msgpacked search fields */
	uint32_t part_count;
};

enum iterator_type {
	/* ITER_EQ must be the first member for request_create  */
	ITER_EQ               =  0, /* key == x ASC order                  */
	ITER_REQ              =  1, /* key == x DESC order                 */
	ITER_ALL              =  2, /* all tuples                          */
	ITER_LT               =  3, /* key <  x                            */
	ITER_LE               =  4, /* key <= x                            */
	ITER_GE               =  5, /* key >= x                            */
	ITER_GT               =  6, /* key >  x                            */
	ITER_BITS_ALL_SET     =  7, /* all bits from x are set in key      */
	ITER_BITS_ANY_SET     =  8, /* at least one x's bit is set         */
	ITER_BITS_ALL_NOT_SET =  9, /* all bits are not set                */
	ITER_OVERLAPS         = 10, /* key overlaps x                      */
	ITER_NEIGHBOR         = 11, /* tuples in distance ascending order from specified point */
	iterator_type_MAX
};

struct tree_iterator {
	struct iterator base;
	const struct memtx_tree *tree;
	struct index_def *index_def;
	struct memtx_tree_iterator tree_iterator;
	enum iterator_type type;
	struct memtx_tree_key_data key_data;
	struct tuple *current_tuple;
	/** Memory pool the iterator was allocated from. */
	struct mempool *pool;
};

]]
)

local index_cont_pairs = function(index, key, last_tuple, opts)
    local gen,param,state = index:pairs(key, opts)
    if not last_tuple then
        return gen,param,state
    end
    local t0 = index:min()
    if not t0 then
        return gen,param,state
    end

    -- TODO: need more checks of last_tuple to be compatible with the space
    -- TODO: check that engine is memtx and index is tree

    local fmt = ffi.C.box_tuple_format(t0)
    local data_start, data_end = box.tuple.encode(last_tuple)
    local tuple = ffi.C.box_tuple_new(fmt, data_start, data_end)
    ffi.C.box_tuple_ref(tuple)

    state = gen(param, state)
    if not (ffi.cast("struct tree_iterator&", state).current_tuple == box.NULL) then
        ffi.C.box_tuple_unref(ffi.cast("struct tree_iterator&", state).current_tuple)
    end
    ffi.cast("struct tree_iterator&", state).current_tuple = tuple

    return gen,param,state
end

local space_cont_pairs = function(space, key, last_tuple, opts)
    box.internal.check_space_arg(space, 'pairs')
    local pk = space.index[0]
    if pk == nil then
        -- empty space without indexes, return empty iterator
        return fun.iter({})
    end
    return index_cont_pairs(pk, key, last_tuple, opts)
end

box.schema.index_mt['cont_pairs'] = index_cont_pairs
box.schema.space_mt['cont_pairs'] = space_cont_pairs
```
</details>

<details>
<summary>In Tarantool 2.6.0-38-gee07eab</summary>

```lua
ffi = require('ffi')

ffi.cdef(
[[
typedef struct tuple_format box_tuple_format_t;
typedef struct tuple box_tuple_t;

box_tuple_format_t *
box_tuple_format(box_tuple_t *tuple);

box_tuple_t *
box_tuple_new(box_tuple_format_t *format, const char *data,
              const char *end);

int
box_tuple_ref(box_tuple_t *tuple);
void
box_tuple_unref(box_tuple_t *tuple);

typedef uint64_t hint_t;

/** @copydoc tuple_compare_with_key() */
typedef int (*tuple_compare_with_key_t)(struct tuple *tuple,
					hint_t tuple_hint,
					const char *key,
					uint32_t part_count,
					hint_t key_hint,
					struct key_def *key_def);
/** @copydoc tuple_compare() */
typedef int (*tuple_compare_t)(struct tuple *tuple_a,
			       hint_t tuple_a_hint,
			       struct tuple *tuple_b,
			       hint_t tuple_b_hint,
			       struct key_def *key_def);
/** @copydoc tuple_extract_key() */
typedef char *(*tuple_extract_key_t)(struct tuple *tuple,
				     struct key_def *key_def,
				     int multikey_idx,
				     uint32_t *key_size);
/** @copydoc tuple_extract_key_raw() */
typedef char *(*tuple_extract_key_raw_t)(const char *data,
					 const char *data_end,
					 struct key_def *key_def,
					 int multikey_idx,
					 uint32_t *key_size);
/** @copydoc tuple_hash() */
typedef uint32_t (*tuple_hash_t)(struct tuple *tuple,
				 struct key_def *key_def);
/** @copydoc key_hash() */
typedef uint32_t (*key_hash_t)(const char *key,
				struct key_def *key_def);
/** @copydoc tuple_hint() */
typedef hint_t (*tuple_hint_t)(struct tuple *tuple,
			       struct key_def *key_def);
/** @copydoc key_hint() */
typedef hint_t (*key_hint_t)(const char *key, uint32_t part_count,
			     struct key_def *key_def);

/* Definition of a multipart key. */
struct key_def {
	/** @see tuple_compare() */
	tuple_compare_t tuple_compare;
	/** @see tuple_compare_with_key() */
	tuple_compare_with_key_t tuple_compare_with_key;
	/** @see tuple_extract_key() */
	tuple_extract_key_t tuple_extract_key;
	/** @see tuple_extract_key_raw() */
	tuple_extract_key_raw_t tuple_extract_key_raw;
	/** @see tuple_hash() */
	tuple_hash_t tuple_hash;
	/** @see key_hash() */
	key_hash_t key_hash;
	/** @see tuple_hint() */
	tuple_hint_t tuple_hint;
	/** @see key_hint() */
	key_hint_t key_hint;
	/**
	 * Minimal part count which always is unique. For example,
	 * if a secondary index is unique, then
	 * unique_part_count == secondary index part count. But if
	 * a secondary index is not unique, then
	 * unique_part_count == part count of a merged key_def.
	 */
	uint32_t unique_part_count;
	/** True, if at least one part can store NULL. */
	bool is_nullable;
	/** True if some key part has JSON path. */
	bool has_json_paths;
	/** True if it is a multikey index definition.
	 * XXX Not used for multikey functional indexes,
	 * please use func->def.is_multikey instead.
	 */
	bool is_multikey;
	/** True if it is a functional index key definition. */
	bool for_func_index;
	/**
	 * True, if some key parts can be absent in a tuple. These
	 * fields assumed to be MP_NIL.
	 */
	bool has_optional_parts;
	/** Key fields mask. @sa column_mask.h for details. */
	uint64_t column_mask;
	/**
	 * A pointer to a functional index function.
	 * Initially set to NULL and is initialized when the
	 * record in _func_index is handled by a respective trigger.
	 * The reason is that we may not yet have a defined
	 * function when a functional index is defined. E.g.
	 * during recovery, we recovery _index first, and _func
	 * second, so when recovering _index no func object is
	 * loaded in the cache and nothing can be assigned.
	 * Once a pointer is assigned its life cycle is guarded by
	 * a check in _func on_replace trigger in alter.cc which
	 * would not let anyone change a function until it is
	 * referenced by a functional index.
	 * In future, one will be able to update a function of
	 * a functional index by disabling the index, thus
	 * clearing this pointer, modifying the function, and
	 * enabling/rebuilding the index.
	 */
	struct func *func_index_func;
	/**
	 * In case of the multikey index, a pointer to the
	 * JSON path string, the path to the root node of
	 * multikey index that contains the array having
	 * index placeholder sign [*].
	 *
	 * This pointer duplicates the JSON path of some key_part.
	 * This path is not 0-terminated. Moreover, it is only
	 * JSON path subpath so key_def::multikey_path_len must
	 * be directly used in all cases.
	 *
	 * This field is not NULL iff this is multikey index
	 * key definition.
	 */
	const char *multikey_path;
	/**
	 * The length of the key_def::multikey_path.
	 * Valid when key_def->is_multikey is true,
	 * undefined otherwise.
	 */
	uint32_t multikey_path_len;
	/**
	 * The index of the root field of the multikey JSON
	 * path index key_def::multikey_path.
	 * Valid when key_def->is_multikey is true,
	 * undefined otherwise.
	*/
	uint32_t multikey_fieldno;
	/** The size of the 'parts' array. */
	uint32_t part_count;
	/** Description of parts of a multipart index. */
	//struct key_part parts[];
};

enum index_type {
	HASH = 0, /* HASH Index */
	TREE,     /* TREE Index */
	BITSET,   /* BITSET Index */
	RTREE,    /* R-Tree Index */
	index_type_MAX,
};

enum rtree_index_distance_type {
	 /* Euclid distance, sqrt(dx*dx + dy*dy) */
	RTREE_INDEX_DISTANCE_TYPE_EUCLID,
	/* Manhattan distance, fabs(dx) + fabs(dy) */
	RTREE_INDEX_DISTANCE_TYPE_MANHATTAN,
	rtree_index_distance_type_MAX
};

/** Index options */
struct index_opts {
	/**
	 * Is this index unique or not - relevant to HASH/TREE
	 * index
	 */
	bool is_unique;
	/**
	 * RTREE index dimension.
	 */
	int64_t dimension;
	/**
	 * RTREE distance type.
	 */
	enum rtree_index_distance_type distance;
	/**
	 * Vinyl index options.
	 */
	int64_t range_size;
	int64_t page_size;
	/**
	 * Maximal number of runs that can be created in a level
	 * of the LSM tree before triggering compaction.
	 */
	int64_t run_count_per_level;
	/**
	 * The LSM tree multiplier. Each subsequent level of
	 * the LSM tree is run_size_ratio times larger than
	 * previous one.
	 */
	double run_size_ratio;
	/* Bloom filter false positive rate. */
	double bloom_fpr;
	/**
	 * LSN from the time of index creation.
	 */
	int64_t lsn;
	/**
	 * SQL specific statistics concerning tuples
	 * distribution for query planer. It is automatically
	 * filled after running ANALYZE command.
	 */
	struct index_stat *stat;
	/** Identifier of the functional index function. */
	uint32_t func_id;
};

struct rlist {
	struct rlist *prev;
	struct rlist *next;
};

/* Definition of an index. */
struct index_def {
	/* A link in key list. */
	struct rlist link;
	/** Ordinal index number in the index array. */
	uint32_t iid;
	/* Space id. */
	uint32_t space_id;
	/** Index name. */
	char *name;
	/** Index type. */
	enum index_type type;
	struct index_opts opts;

	/* Index key definition. */
	struct key_def *key_def;
	/**
	 * User-defined key definition, merged with the primary
	 * key parts. Used by non-unique keys to uniquely identify
	 * iterator position.
	 */
	struct key_def *cmp_def;
};

struct index {
	/** Virtual function table. */
	const struct index_vtab *vtab;
	/** Engine used by this index. */
	struct engine *engine;
	/* Description of a possibly multipart key. */
	struct index_def *def;
	/** Reference counter. */
	int refs;
	/* Space cache version at the time of construction. */
	uint32_t space_cache_version;
};

struct iterator {
    /**
     * Iterate to the next tuple.
     * The tuple is returned in @ret (NULL if EOF).
     * Returns 0 on success, -1 on error.
     */
    int (*next)(struct iterator *it, struct tuple **ret);
    /** Destroy the iterator. */
    void (*free)(struct iterator *);
    /** Space cache version at the time of the last index lookup. */
    uint32_t space_cache_version;
    /** ID of the space the iterator is for. */
    uint32_t space_id;
    /** ID of the index the iterator is for. */
    uint32_t index_id;
    /**
     * Pointer to the index the iterator is for.
     * Guaranteed to be valid only if the schema
     * state has not changed since the last lookup.
     */
    struct index *index;
};

struct matras_view {
    /* root extent of the view */
    void *root;
    /* block count in the view */
    uint32_t block_count;
    /* all views are linked into doubly linked list */
    struct matras_view *prev_view, *next_view;
};

struct memtx_tree_iterator {
    /* ID of a block, containing element. -1 for an invalid iterator */
    uint32_t block_id;
    /* Position of an element in the block. Could be -1 for last in block*/
    uint16_t pos;
    /* Version of matras memory for MVCC */
    struct matras_view view;
};

struct memtx_tree_key_data {
    /** Sequence of msgpacked search fields. */
    const char *key;
    /** Number of msgpacked search fields. */
    uint32_t part_count;
    /** Comparison hint, see tuple_hint(). */
    hint_t hint;
};

struct memtx_tree_data {
    /* Tuple that this node is represents. */
    struct tuple *tuple;
    /** Comparison hint, see key_hint(). */
    hint_t hint;
};

enum iterator_type {
    /* ITER_EQ must be the first member for request_create  */
    ITER_EQ               =  0, /* key == x ASC order                  */
    ITER_REQ              =  1, /* key == x DESC order                 */
    ITER_ALL              =  2, /* all tuples                          */
    ITER_LT               =  3, /* key <  x                            */
    ITER_LE               =  4, /* key <= x                            */
    ITER_GE               =  5, /* key >= x                            */
    ITER_GT               =  6, /* key >  x                            */
    ITER_BITS_ALL_SET     =  7, /* all bits from x are set in key      */
    ITER_BITS_ANY_SET     =  8, /* at least one x's bit is set         */
    ITER_BITS_ALL_NOT_SET =  9, /* all bits are not set                */
    ITER_OVERLAPS         = 10, /* key overlaps x                      */
    ITER_NEIGHBOR         = 11, /* tuples in distance ascending order from specified point */
    iterator_type_MAX
};

struct tree_iterator {
    struct iterator base;
    struct memtx_tree_iterator tree_iterator;
    enum iterator_type type;
    struct memtx_tree_key_data key_data;
    struct memtx_tree_data current;
    /** Memory pool the iterator was allocated from. */
    struct mempool *pool;
};

]]
)

local index_cont_pairs = function(index, key, last_tuple, opts)
    local gen,param,state = index:pairs(key, opts)
    if not last_tuple then
        return gen,param,state
    end
    local t0 = index:min()
    if not t0 then
        return gen,param,state
    end

    -- TODO: need more checks of last_tuple to be compatible with the space
    -- TODO: check that engine is memtx and index is tree

    local fmt = ffi.C.box_tuple_format(t0)
    local data_start, data_end = box.internal.tuple.encode(last_tuple)
    local tuple = ffi.C.box_tuple_new(fmt, data_start, data_end)
    ffi.C.box_tuple_ref(tuple)

    state = gen(param, state)
    if not (ffi.cast("struct tree_iterator&", state).current.tuple == box.NULL) then
        ffi.C.box_tuple_unref(ffi.cast("struct tree_iterator&", state).current.tuple)
    end
    ffi.cast("struct tree_iterator&", state).current.tuple = tuple
    local def = ffi.cast("struct tree_iterator&", state).base.index.def.cmp_def
    local hint = def.tuple_hint(tuple, def)
    ffi.cast("struct tree_iterator&", state).current.hint = hint

    return gen,param,state
end

local space_cont_pairs = function(space, key, last_tuple, opts)
    box.internal.check_space_arg(space, 'pairs')
    local pk = space.index[0]
    if pk == nil then
        -- empty space without indexes, return empty iterator
        return fun.iter({})
    end
    return index_cont_pairs(pk, key, last_tuple, opts)
end

box.schema.index_mt['cont_pairs'] = index_cont_pairs
box.schema.space_mt['cont_pairs'] = space_cont_pairs
```
</details>

You will get a new method `cont_pairs` in index and space, in which you can specify tuple from which you want to continue iteration.

<details>
<summary>Example</summary>

```lua

tarantool> s = box.schema.space.create('test', {engine = 'memtx'})
tarantool> i1 = s:create_index('test1', {type='tree', parts={{1, 'uint'}}})
tarantool> i2 = s:create_index('test2', {type='tree', parts={{2, 'uint'}}, unique=false})
tarantool> box.schema.user.grant('guest', 'read,write', 'space', 'test')
tarantool> for i = 1,3 do for j = 0,9 do s:replace{i * 10 + j, i} end end
...

tarantool> for _,v in i2:cont_pairs({2}, {22, 2}) do print(v) end
[23, 2]
[24, 2]
[25, 2]
[26, 2]
[27, 2]
[28, 2]
[29, 2]
---
...

tarantool> for _,v in i2:cont_pairs({2}, {27, 2}) do print(v) end
[28, 2]
[29, 2]
---
...

tarantool> for _,v in i2:cont_pairs({2}, {27, 2}, {iterator='REQ'}) do print(v) end
[26, 2]
[25, 2]
[24, 2]
[23, 2]
[22, 2]
[21, 2]
[20, 2]
---
...

tarantool> for _,v in i2:cont_pairs({}, {27, 2}) do print(v) end
[28, 2]
[29, 2]
[30, 3]
[31, 3]
[32, 3]
[33, 3]
[34, 3]
[35, 3]
[36, 3]
[37, 3]
[38, 3]
[39, 3]
---
...

tarantool> for _,v in s:cont_pairs({}, {27, 2}) do print(v) end
[28, 2]
[29, 2]
[30, 3]
[31, 3]
[32, 3]
[33, 3]
[34, 3]
[35, 3]
[36, 3]
[37, 3]
[38, 3]
[39, 3]
---
...

tarantool> for _,v in s:cont_pairs({}, {27, 2}, {iterator='LE') do print(v) end
---
- error: '[string "for _,v in s:cont_pairs({}, {27, 2}, {iterato..."]:1: ''}'' expected
    near '')'''
...

tarantool> for _,v in s:cont_pairs({}, {27, 2}, {iterator='LE'}) do print(v) end
[26, 2]
[25, 2]
[24, 2]
[23, 2]
[22, 2]
[21, 2]
[20, 2]
[19, 1]
[18, 1]
[17, 1]
[16, 1]
[15, 1]
[14, 1]
[13, 1]
[12, 1]
[11, 1]
[10, 1]
---
...

tarantool> for _,v in s:cont_pairs({}, {22, 2}, {iterator='LE'}) do print(v) end
[21, 2]
[20, 2]
[19, 1]
[18, 1]
[17, 1]
[16, 1]
[15, 1]
[14, 1]
[13, 1]
[12, 1]
[11, 1]
[10, 1]
---
...
```
</details>

That only works for memtx now, but I think it's possible to implement the same for vinyl.
I'm sure that one of those two implementation will work on any tarantool version starting from 1.10.

**Comment by alyapunov** [comment]
The hack above is the only option for 1.10 and a temporary solution for 2.6.
If that API is suitable, we will implement an option `from_tuple` among `select()` and `pairs()` options (for 2.x, but not for 1.10).

**Comment by Mons** [comment]
I probose a bit another naming with the following cases:

1. Usecase for implemented in future:

* use `after` to indicate "after this tuple"
* use `after_key` to indicate "after this value of key in this `unique` index"
* use `after_primary` (or `after_pk`) to indicate "after this value of key in primary index" (for non-unique indexes)

Example:
```
sp:select({key}, { limit = 10, after => { ...tuple } })
sp:pairs({key}, { iterator = 'EQ', after => { ...tuple } })
```


2. For current workaround name method to point to the future: `pairs_after`, `pairs_after_key`, `pairs_after_pk`

Example:
```
tarantool> for _,v in i2:pairs_after({2}, {27, 2}) do print(v) end
[28, 2]
[29, 2]

tarantool> for _,v in i2:pairs_after_pk({2}, {27}) do print(v) end
[28, 2]
[29, 2]

```

**Comment by alyapunov** [comment]
After offline discussion I consider that issue to be solved for both 1.10 and 2.5. I also moved `after` option proposal to #5282.

**Comment by olegrok** [comment]
```lua
state = gen(param, state)
if not (ffi.cast("struct tree_iterator&", state).current.tuple == box.NULL) then
    ffi.C.box_tuple_unref(ffi.cast("struct tree_iterator&", state).current.tuple)
end
```

This code will fail with segmentation fault for some cases.
Naive solution:
```lua
local t
state, t = gen(param, state)
if t == nil then
    return fun.iter({})
end

if not (ffi.cast("struct tree_iterator&", state).current.tuple == box.NULL) then
    ffi.C.box_tuple_unref(ffi.cast("struct tree_iterator&", state).current.tuple)
end
```
